### PR TITLE
#17601 A new validation was added to the check in process to verify if the ES cluster is in read only mode

### PR DESCRIPTION
**Pull request by nollymar**

### CODE REVIEW

**Comment by dsilvam** [comment]
Needs integration test. In order to test, the `new ESIndexAPI()` dependency in ESContentletAPIImpl needs to be passed via constructor or setter to be able to provide a Mock version that returns true for `.isClusterInReadOnlyMode()`

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 1]
```diff
@@ -3648,6 +3648,11 @@ private Contentlet internalCheckin(Contentlet contentlet,
             }
         }
 
+        if (!isCheckInSafe(contentRelationships)){
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: needs integration test

**Review by dsilvam — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java (hunk 1 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: Done

**Comment by dsilvam** [comment]
@nollymar tests look good, although I think we are missing a test trying to checking in content with the read-only mode. Reason is If some refactor happens and removes the call of isCheckInSafe we wouldn't have any tests failing.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java [hunk 2]
```diff
@@ -1329,4 +1329,22 @@ Map performLowLevelRequest(Request request) {
 				.readValue(response.getEntity().getContent(), mappingClass));
 	}
 
+    /**
+     *
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java]: this should have doc

Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java]: Done

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 3]
```diff
@@ -159,5 +179,146 @@ public void testGetRelatedContentFromMultilingualContent()
         }
     }
 
+    @Test(expected= DotContentletStateException.class)
+    public void testCheckInWithLegacyRelationshipsAndReadOnlyClusterShouldThrowAnException()
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: doc?, The name is good but I think it should has doc anyway

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 4]
```diff
@@ -159,5 +179,146 @@ public void testGetRelatedContentFromMultilingualContent()
...
+        }finally{
+            if (contentType != null && contentType.id() != null){
+                contentTypeAPI.delete(contentType);
+            }
+        }
+    }
+
+    @Test
+    public void testIsCheckInSafeWithLegacyRelationshipsShouldReturnFalse()
+            throws DotDataException, DotSecurityException {
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: doc?

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 5]
```diff
@@ -159,5 +179,146 @@ public void testGetRelatedContentFromMultilingualContent()
...
+
+        }finally{
+            if (contentType != null && contentType.id() != null){
+                contentTypeAPI.delete(contentType);
+            }
+        }
+    }
+
+    @Test
+    public void testIsCheckInSafeWithRelationshipsFieldsShouldReturnTrue()
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: doc?

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 6]
```diff
@@ -159,5 +179,146 @@ public void testGetRelatedContentFromMultilingualContent()
...
+
+        }finally{
+            if (contentType != null && contentType.id() != null){
+                contentTypeAPI.delete(contentType);
+            }
+        }
+    }
+
+    @Test
+    public void testIsCheckInSafeWithoutRelationshipsShouldReturnTrue() {
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: doc?

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java [hunk 7]
```diff
@@ -159,5 +179,146 @@ public void testGetRelatedContentFromMultilingualContent()
...
+        assertTrue(
+                new ESContentletAPIImpl().isCheckInSafe(null, esIndexAPI));
+    }
+
+
+    private ContentType createContentType(final String name) throws DotSecurityException, DotDataException {
+        return contentTypeAPI.save(ContentTypeBuilder.builder(SimpleContentType.class).folder(
+                FolderAPI.SYSTEM_FOLDER).host(Host.SYSTEM_HOST).name(name)
+                .owner(user.getUserId()).build());
+    }
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: I think are we missing a test: testCheckInWithoutLegacyRelationshipsAndReadOnlyClusterShouldThrowAnException?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 8]
```diff
@@ -4397,6 +4404,42 @@ else if (incomingFile.exists() ){
...
+     * @return
+     */
+    @VisibleForTesting
+    boolean isCheckInSafe(final ContentletRelationships relationships, final ESIndexAPI esIndexAPI) {
+
+        if (relationships != null && relationships.getRelationshipsRecords().size() > 0) {
+            boolean isClusterReadOnly = esIndexAPI.isClusterInReadOnlyMode();
+
+            if (isClusterReadOnly) {
+                for (ContentletRelationships.ContentletRelationshipRecords records : relationships
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: maybe this code should go to a new method

```suggestion
boolean hasLegacyRelationships(...) {
                for (ContentletRelationships.ContentletRelationshipRecords records : relationships
                                              .getRelationshipsRecords()) {
                    if (!records.getRelationship().isRelationshipField()){
                        return true;
                    }
                }
                
                return false;
}                
```

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java [hunk 9]
```diff
@@ -1329,4 +1329,23 @@ Map performLowLevelRequest(Request request) {
...
+                    .getClient().cluster()
+                    .getSettings(new ClusterGetSettingsRequest(), RequestOptions.DEFAULT);
+
+            return Boolean.valueOf(response.getSetting("cluster.blocks.read_only"));
+        } catch (IOException e) {
+            Logger.warnAndDebug(ESIndexAPI.class, "Error getting ES cluster settings", e);
+        }
+
+        return true;
+    }
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java]: test for this method?

**Review by freddyucv — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java (hunk 7 above)
Comment (nollymar) [reply dotCMS/src/integration-test/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImplTest.java]: For this case we shouldn't throw an exception

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java (hunk 9 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESIndexAPI.java]: already exists

**Review by freddyucv — APPROVED** [review]

**Pull request merged**